File: main/client/src/mill/main/client/lock/FileLock.java
package mill.main.client.lock;

import java.io.RandomAccessFile;
import java.nio.channels.FileChannel;

class FileLock extends Lock {

  private final RandomAccessFile raf;
  private final FileChannel chan;

  public FileLock(String path) throws Exception {
    raf = new RandomAccessFile(path, "rw");
    chan = raf.getChannel();
  }

  public Locked lock() throws Exception {
    return new FileLocked(chan.lock());
  }

  public TryLocked tryLock() throws Exception {
    return new FileTryLocked(chan.tryLock());
  }

  public boolean probe() throws Exception {
    java.nio.channels.FileLock l = chan.tryLock();
    if (l == null) return false;
    else {
      l.release();
      return true;
    }
  }

  @Override
  public void close() throws Exception {
    chan.close();
    raf.close();
  }

  public void delete() throws Exception {
    close();
  }
}

File: main/client/src/mill/main/client/lock/FileLock.scala
package mill.main.client.lock

import java.io.RandomAccessFile
import java.nio.channels.FileChannel
import scala.util.Try

class FileLock(path: String) extends Lock with AutoCloseable {
  private final val raf: RandomAccessFile = new RandomAccessFile(path, "rw")
  private final val chan: FileChannel = raf.getChannel

  @throws[Exception]
  def lock(): Locked = new FileLocked(chan.lock())

  @throws[Exception]
  def tryLock(): TryLocked = new FileTryLocked(chan.tryLock())

  @throws[Exception]
  def probe(): Boolean = {
    val lock = chan.tryLock()
    if (lock == null) false
    else {
      lock.release()
      true
    }
  }

  @throws[Exception]
  override def close(): Unit = {
    chan.close()
    raf.close()
  }

  @throws[Exception]
  override def delete(): Unit =
    close()
}
